src/sparsezoo/v2/utils/authentication.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import errno
+import json
+import logging
+import os
+import time
+from datetime import datetime, timezone
+from typing import Dict
+
+import requests
+import yaml
+
+
+BASE_API_URL = (
+    os.getenv("SPARSEZOO_API_URL")
+    if os.getenv("SPARSEZOO_API_URL")
+    else "https://api.neuralmagic.com"
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PUBLIC_AUTH_TYPE = "public"
+CREDENTIALS_YAML_TOKEN_KEY = "nm_api_token"
+AUTH_API = f"{BASE_API_URL}/auth"
+NM_TOKEN_HEADER = "nm-token-header"
+
+
+def create_dirs(path: str):
+    """
+    :param path: the directory path to try and create
+    """
+    path = clean_path(path)
+
+    try:
+        os.makedirs(path)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            pass
+        else:
+            # Unexpected OSError, re-raise.
+            raise
+
+
+def create_parent_dirs(path: str):
+    """
+    :param path: the file path to try to create the parent directories for
+    """
+    parent = os.path.dirname(path)
+    create_dirs(parent)
+
+
+def clean_path(path: str) -> str:
+    """
+    :param path: the directory or file path to clean
+    :return: a cleaned version that expands the user path and creates an absolute path
+    """
+    return os.path.abspath(os.path.expanduser(path))
+
+
+CREDENTIALS_YAML = os.path.abspath(
+    os.getenv("SPARSEZOO_CREDENTIALS")
+    if os.getenv("SPARSEZOO_CREDENTIALS")
+    else clean_path(os.path.join("~", ".cache", "sparsezoo", "credentials.yaml"))
+)
+
+
+class SparseZooCredentials:
```

Comment:
I think this could be reworked into two private_functions:
1. _save_token (unchanged from the method, but just a standalone function)
2. _get_token

The `_get_token` function could combine the `__init__` function and the `token` property so all the invariants are in one place.

Something like:

```python
def _get_token():
    if not os.path.exists(CREDENTIALS_YAML):
        return None
    with open(...) as fp:
        creds = yaml.safe_load(...)
    if creds is None or TOKEN_KEY not in creds:
       return None
    token = creds[TOKEN_KEY]["token"]
    created = creds[TOKEN_KEY]["created"]

    creation_difference = ...
    if creation_difference > 30:
        return None

    return token
```

then in get_auth_header you could just call `_get_token` in place of instantiating this class and access the token property

Reply:
Good catch. Just for context, there is a bunch of functions that I am directly copying-pasting from sparsezoo v1; this one included. Once again, I propose to leave those as they are for now, and do a proper boy-scout refactoring before they land in main. Open for discussion.